dask/array/random.py

```diff
@@ -98,40 +99,43 @@ def _broadcast_any(ar, shape, chunks):
         dsk = {}
         dsks = []
         lookup = {}
-        small_args = []
-        for i, ar in enumerate(args):
-            if isinstance(ar, (np.ndarray, Array)):
-                res = _broadcast_any(ar, size, chunks)
-                if isinstance(res, Array):
-                    dsks.append(res.dask)
-                    lookup[i] = res.name
-                elif isinstance(res, np.ndarray):
-                    name = 'array-{}'.format(tokenize(res))
-                    lookup[i] = name
-                    dsk[name] = res
-                small_args.append(ar[tuple(0 for _ in ar.shape)])
-            else:
-                small_args.append(ar)
-
-        small_kwargs = {}
-        for key, ar in kwargs.items():
-            if isinstance(ar, (np.ndarray, Array)):
-                res = _broadcast_any(ar, size, chunks)
-                if isinstance(res, Array):
-                    dsks.append(res.dask)
-                    lookup[key] = res.name
-                elif isinstance(res, np.ndarray):
-                    name = 'array-{}'.format(tokenize(res))
-                    lookup[key] = name
-                    dsk[name] = res
-                small_kwargs[key] = ar[tuple(0 for _ in ar.shape)]
-            else:
-                small_kwargs[key] = ar
 
-        # Get dtype
-        small_kwargs['size'] = (0,)
         func = getattr(np.random.RandomState(), funcname)
-        dtype = func(*small_args, **small_kwargs).dtype
+        if funcname == 'multivariate_normal':
+            dtype = func(*args, size=(0,), **kwargs).dtype
```

Comment:
This special-casing seems odd to me.  It looks like the code below is designed to catch dask inputs to the da.random functions.  This branch seems to avoid that entirely.  

What happens if someone passes a dask array to covariance or mean?
Is there some way to continue down this code path?

Alternatively, if we do have to special-case multivariate-normal (which is fine as long as we err informatively) then would it be possible to do that special casing outside of the core function?  It feels off to include a use of this function inside the main body of the function.

Reply:
I wasn't really sure what the code was doing or why it was doing it so I short-circuited it because from what I can see I don't think it's appropriate as-is for `multivariate_normal`.

I just tested and it appears to work fine dask arrays for mean/cov. I assume they just get converted to an ndarray when/where they're used in which case it might be exactly what you'd want?

Reply:
I guess let me back up and say "why is it necessary to special case things here?"  This seems to me like the kind of things that will require extra maintenance in the future, so I'd like to avoid it if possible.

Reply:
I discuss that in more detail in my other comment below which boils down to:
 1. I don't think the broadcasting behaviour is correct for `multivariate_normal` when the underlying numpy function doesn't support it
 2. The implementation of `small_{args|kwargs}` doesn't work for `multivariate_normal`

AFAICS the simplest thing to do is to just skip this for `multivariate_normal`. Absent this I think the implementation would need a much bigger re-write that I don't think I'm capable of as I not familiar enough with dask internals.

Maybe there *is* a much neater/cleaner way to implement this but I'm not sure how. I'm totally fine with you closing this PR and doing it properly - I'm not wedded to this code and agree the special-casing isn't great; I just can't see any easier/cleaner way myself.

Reply:
> I'm totally fine with you closing this PR and doing it properly 

Heh, I think that you overestimate the amount of free time I have :)

Also, I don't know how to do this properly.  I would need to dive in to figure out what things are doing.  The thing to do is to understand what the previous code was trying to accomplish and then make sure that we're not breaking some implicit contract.  Some approaches here include using git blame to find the commit that introduce this code and read about the commit message or look at tests that were added at the same time.

It *looks* like it's trying to include the graphs of dask-y inputs into the resulting random array (rather than computing immediately and including the numpy arrays).

Reply:
I can try to get around to doing this properly as I do need to get a better understanding of the dask internals but I can't say when I might find the time. This isn't on my critical-path since there's a fairly easy work-around but I'm happy to let this lie around until I can dive into it properly...